advanced/functions/server/serverfunction.ts

```diff
@@ -0,0 +1,104 @@
+import {PackRef} from "../../ts-classes/packRef";
+// import {PackWeight} from "../../ts-classes/packWeight";
+import {YPackWeight} from "sample-invoicing/advanced";
+import {Pack} from "../../ts-classes/pack";
+import {Scales} from "../common/commonfunction";
+
+type ServerFunctionResponse<T> = Promise<{data:T}>
+
+exports.REST = {
+    $add: {
+        module: "advanced",
+        name: "add",
+        scope: "public",
+        description: "add a new package with given title and description",
+        restricted: false,
+        args: [],
+        content: [
+            {
+                name: "title",
+                type: "string",
+                mandatory: true,
+                description: "title of the Pack"
+            },
+            {
+                name: "modelRef",
+                type: "string",
+                mandatory: true,
+                description: "ref to the model"
+            },
+            {
+                name: "weight",
+                type: "number",
+                mandatory: true,
+                description: "weight value"
+            },
+            {
+                name: "unit",
+                type: "number",
+                mandatory: true,
+                description: "weight unit"
+            },
+            {
+                name: "similar",
+                type: "array",
+                description: "refs to similar packs"
+            }
+        ],
+        output: "{success: true, mass: 10}",
+        example: "/advanced/serverfunction/add",
+        method: "post"
+    },
+    add: async (
+        _args: any,
+        _content: { title: string, modelRef: string, weight: number, unit: number, similar?: string[]},
+        _serv: any,
+        _role: YRole,
+        _session: YSession
+    ): ServerFunctionResponse<{
+        success: boolean, mass: number
+    }> => {
+        const [newPack, modelRef]:[PackUpsale, Model] = await Promise.all([
+            _role.advanced.createClass("packUpsale"),
+            _role.simple.createClass("model")
+        ]);
+        newPack.title = _content.title;
+        modelRef.objectId = _content.modelRef;
+        newPack.model = modelRef as YTref<Model>;
+        newPack.weight.value = _content.weight;
+        newPack.weight.unit = _content.unit;
+
+        await Promise.all((_content.similar || []).map(async (packId) => {
+            const [newRef, packRef]:[PackRef, PackUpsale] = await Promise.all([
+                _role.advanced.createClass("packRef"),
+                _role.advanced.createClass("packUpsale")
+            ]);
+            packRef.objectId = packId;
+            newRef.pack = packRef as YTref<Pack>;
+            return newPack.similar.push(newRef);
+        }));
+
+        newPack.validate(err => {
+            if (err) {
+                throw err;
+            }
+        }, false);
+
+        const cnx = await _session.dbserver.connect("simple", 'simple', null);
+        await cnx.commit([newPack]);
```

Comment:
This gives me 
```
{ YSQLError: YSQLError: YPlatform/DatabaseStore - connection: simple, error received from the DB: 
Error: Invalid field value : Missing mandatory slot : advanced.pack.weight_value
     at new YFWKError (/app/yupana/framework/server/lib/core/YError.js:15:9)
     at new YSQLError (/app/yupana/framework/server/lib/core/YError.js:22:9)
     at Object.yaddon.storeDatabase (/app/yupana/service/common/YQueryAddon.js:177:25) HTTP: undefined }
```

Reply:
Can you try by doing ```newPack.weight_value = _content.weight;``` with underscore between ```weight``` and ```value```, just to confirm if it's a bug of YClass.

Reply:
yep, it does the trick if I patch compiled js.

Reply:
Are you in the known bug of YClass ?
https://github.com/YupanaInc/Yupana-Framework/blob/build/2.5.0/server/test/integration/services/integration-master/tests/test-yclass.js#L203
Normally the complex types works, except when in groups. So I don't know why your example doesn't...

Reply:
but it seems that it's the case that it works in the sense that the object property is set correctly (like in the test), but later when it tries to commit it fails because the commit uses the result of `toJSON()` which only outputs the properties that are in the updated slots list — and setting the value with the dot notation doesn't do that...

is there a test that sets the property using dots and then commits the object?
otherwise, this use-case isn't covered in the tests...  :)

Reply:
We have some test on updatedSlots which is internally used by commit.